File: neetcode/blind75/Tries/design_word_search_data_structure.py
# https://neetcode.io/problems/design-word-search-data-structure

class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        trie_node = self.getRoot()

        for c in word:
            index = ord(c) - ord('a')

            if (trie_node.children[index] is None):
                trie_node.children[index] = TrieNode()

            trie_node = trie_node.children[index]

        trie_node.isEnd = True

    def search(self, word: str) -> bool:
        trie_node = self.getRoot()

        def dfs(node, word, w_index):
            if node is None:
                # no match
                return False
            elif w_index == len(word):
                # at end of word, check if marked end
                return node.isEnd
            elif self.isEmpty(node):
                # trie links terminated before end of input word
                return False
            
            if (word[w_index] == '.'):
                # if . then must go into each character
                for c in range(len(node.children)):
                    if (node.children[c] is not None):
                        if dfs(node.children[c], word, w_index + 1):
                            # if ret True, then it means path to found word exists. Propagate up
                            return True

                return False
            else:
                index = ord(word[w_index]) - ord('a')
                return dfs(node.children[index], word, w_index + 1)

        return dfs(trie_node, word, 0)

    def getRoot(self):
        return self.root

    def isEmpty(self, node):
        for c in node.children:
            if c is not None:
                return False

        return True
        

File: neetcode/blind75_JS/Tries/design-word-search-data-structure.js
// https://neetcode.io/problems/design-word-search-data-structure

class TrieNode {
    constructor() {
        this.children = Array(26).fill(null)
        this.isEnd = false
    }
}

class WordDictionary {
    constructor() {
        this.root = new TrieNode()
    }

    /**
     * @param {string} word
     * @return {void}
     */
    addWord(word) {
        let currNode = this.root

        for (let c of word) {
            const ord = c.charCodeAt(0) - 'a'.charCodeAt(0)
            if (currNode.children[ord] === null) {
                currNode.children[ord] = new TrieNode()
            }

            currNode = currNode.children[ord]
        }

        currNode.isEnd = true
    }

    /**
     * @param {string} word
     * @return {boolean}
     */
    search(word) {
        return this.searchDFS(word, this.root)
    }

    searchDFS(word, currNode) {
        if (currNode === null) {
            return false
        }
        if (word.length === 0) {
            return currNode.isEnd
        }
        
        if (word[0] === '.'){
            // must check all children that are not null
            for (let c of currNode.children) {
                if (this.searchDFS(word.slice(1), c)) {
                    return true
                }
            }
        } else {
            const ord = word[0].charCodeAt(0) - 'a'.charCodeAt(0)
            if (this.searchDFS(word.slice(1), currNode.children[ord])) {
                return true
            }
        }

        return false
    }
}

